contracts/wrapped-polyx/lib.rs

```diff
@@ -0,0 +1,456 @@
+        },
+        Mapping,
+    };
+    use alloc::vec;
+
+    use crate::*;
+
+    /// A contract to wrap and unwrap POLYX (native token)
+    #[ink(storage)]
+    #[derive(Default, SpreadAllocate)]
+    pub struct WrappedPolyx {
+      ticker: Ticker,
+      initialized: bool,
+      /// Venue for settlements.
+      venue: VenueId,
+      /// Contract's identity.
+      did: IdentityId,
+      /// Custodial portfolios.
+      portfolios: Mapping<IdentityId, PortfolioId>,
+    }
+
+    /// Event emitted when POLYX is wrapped
+    #[ink(event)]
+    pub struct PolyxWrapped {
+      #[ink(topic)]
+      did: IdentityId,
+      #[ink(topic)]
+      key: AccountId,
+      amount: Balance,
+    }
+
+    /// Event emitted when POLYX is unwrapped
+    #[ink(event)]
+    pub struct PolyxUnwrapped {
+      #[ink(topic)]
+      did: IdentityId,
+      #[ink(topic)]
+      key: AccountId,
+      amount: Balance,
+    }
+
+    /// Event emitted when POLYX is unwrapped
+    #[ink(event)]
+    pub struct PortfolioAdded {
+      #[ink(topic)]
+      did: IdentityId,
+      portfolio_kind: PortfolioKind,
+    }
+
+    /// Event emitted when POLYX is unwrapped
+    #[ink(event)]
+    pub struct PortfolioRemoved {
+      #[ink(topic)]
+      did: IdentityId,
+      portfolio_kind: PortfolioKind,
+    }
+
+    /// The contract error types.
+    #[derive(Debug, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        /// Polymesh runtime error.
+        PolymeshError(PolymeshError),
+        /// Scale decode failed.
+        ScaleError,
+        /// Missing Identity.  MultiSig's are not supported.
+        MissingIdentity,
+        /// Contract hasn't been initialized.
+        NotInitialized,
+        /// Contract has already been initialized.
+        AlreadyInitialized,
+        /// Invalid portfolio authorization.
+        InvalidPortfolioAuthorization,
+        /// The caller has already initialized a portfolio.
+        AlreadyHavePortfolio,
+        /// The caller doesn't have a portfolio yet.
+        NoPortfolio,
+    }
+
+    impl From<PolymeshError> for Error {
+      fn from(err: PolymeshError) -> Self {
+        Self::PolymeshError(err)
+      }
+    }
+
+    impl From<PolymeshRuntimeErr> for Error {
+      fn from(err: PolymeshRuntimeErr) -> Self {
+        Self::PolymeshError(err.into())
+      }
+    }
+
+    /// The contract result type.
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    impl WrappedPolyx {
+        /// Creates a new contract.
+        #[ink(constructor)]
+        pub fn new(ticker: Ticker) -> Self {
+          ink_lang::utils::initialize_contract(|contract| {
+              Self::new_init(contract, ticker)
+          })
+        }
+
+        fn new_init(&mut self, ticker: Ticker) {
+          // The contract should always have an identity.
+          self.did = self.get_did(Self::env().account_id()).unwrap();
+          self.ticker = ticker;
```

Comment:
`self.ticker` can be default initialized, then set to the correct value when `init` is called.